crypto/src/memo.rs

```diff
@@ -16,12 +23,107 @@ impl Default for MemoPlaintext {
 }
 
 impl MemoPlaintext {
-    pub fn encrypt(&self, _key: &Address) -> MemoCiphertext {
-        // TODO!
-        let memo_bytes = [0u8; MEMO_CIPHERTEXT_LEN_BYTES];
-        MemoCiphertext(memo_bytes)
+    /// Encrypt a memo, returning its ciphertext.
+    pub fn encrypt(
+        &self,
+        esk: &ka::Secret,
+        transmission_key: &ka::Public,
+        diversified_generator: &decaf377::Element,
+    ) -> Result<MemoCiphertext, anyhow::Error> {
+        let epk = esk.diversified_public(diversified_generator);
+        let shared_secret = esk
+            .key_agreement_with(&transmission_key)
+            .map_err(|_| anyhow!("could not perform key agreement"))?;
+
+        // Use Blake2b-256 to derive encryption key.
+        let mut kdf_params = blake2b_simd::Params::new();
+        kdf_params.hash_length(32);
+        let mut kdf = kdf_params.to_state();
+        kdf.update(&shared_secret.0);
+        kdf.update(&epk.0);
+        let kdf_output = kdf.finalize();
+        let key = Key::from_slice(kdf_output.as_bytes());
+
+        let cipher = ChaCha20Poly1305::new(key);
+        let nonce = Nonce::from_slice(&[1u8; 12]);
+
+        let encryption_result = cipher
+            .encrypt(nonce, self.0.as_ref())
+            .map_err(|_| anyhow!("encryption error!"))?;
+
+        let ciphertext: [u8; MEMO_CIPHERTEXT_LEN_BYTES] = encryption_result
+            .try_into()
+            .map_err(|_| anyhow!("memo encryption result does not fit in ciphertext len"))?;
```

Comment:
When would this happen? We know the size of the memo plaintext, so we know the size of the ciphertext, right?

Could this be an `.expect` instead?